t/04-openresty/ffi/304-proxy_wasm_set_property_getter.t

```diff
@@ -0,0 +1,597 @@
+# vim:set ft= ts=4 sw=4 et fdm=marker:
+
+use strict;
+use lib '.';
+use t::TestWasm::Lua;
+
+skip_no_openresty();
+
+plan tests => repeat_each() * (blocks() * 5);
+
+run_tests();
+
+__DATA__
+
+=== TEST 1: set_property_getter() - setting host properties getter fails on init_worker_by_lua
+--- wasm_modules: on_phases
+--- http_config
+    init_worker_by_lua_block {
+        local proxy_wasm = require "resty.wasmx.proxy_wasm"
+        local filters = {
+            { name = "on_phases" },
+        }
+
+        _G.c_plan = assert(proxy_wasm.new(filters))
+        assert(proxy_wasm.load(_G.c_plan))
+
+        local function getter(key)
+            if key == "my_key" then
+                return true, "my_value"
+            end
+            return false
+        end
+
+        assert(proxy_wasm.set_property_getter(getter))
+    }
+--- config
+    location /t {
+        rewrite_by_lua_block {
+            ngx.say("ok")
+        }
+    }
+--- response_body
+ok
+--- error_log eval
+[
+    qr/\[error\] .*? cannot set host properties getter outside of a request/,
+    qr/\[error\] .*? init_worker_by_lua(\(nginx\.conf:\d+\))?:\d+: could not set property getter/,
+]
+--- no_error_log
+[crit]
+
+
+
+=== TEST 2: set_property_getter() - setting host properties getter works on rewrite_by_lua
```

Comment:
I would like to understand what happens in this test because it seems wrong, but I don't. We ask for a log of `wasmx.my_property` in `on_response_body` (which is called twice), and we get:

```
2023/11/21 17:14:07 [debug] 1992222#0: *2 [proxy-wasm]["hostcalls" #1] filter 1/1 resuming "on_response_body" step in "body_filter" phase                                                                         
2023/11/21 17:14:07 [info] 1992222#0: *2 [proxy-wasm]["hostcalls" #1] [hostcalls] on_response_body, 3 bytes, eof: false, client: 127.0.0.1, server: localhost, request: "GET /t HTTP/1.1", host: "localhost"      │
2023/11/21 17:14:07 [info] 1992222#0: *2 [proxy-wasm]["hostcalls" #1] [hostcalls] testing in "ResponseBody", client: 127.0.0.1, server: localhost, request: "GET /t HTTP/1.1", host: "localhost"
...                                                 
2023/11/21 17:14:07 [info] 1992222#0: *2 [proxy-wasm]["hostcalls" #1] wasmx.my_property: my_value, client: 127.0.0.1, server: localhost, request: "GET /t HTTP/1.1", host: "localhost"                            
...
2023/11/21 17:14:07 [debug] 1992222#0: *2 [proxy-wasm]["hostcalls" #1] filter 1/1 resuming "on_response_body" step in "body_filter" phase                                                                         
2023/11/21 17:14:07 [info] 1992222#0: *2 [proxy-wasm]["hostcalls" #1] [hostcalls] on_response_body, 0 bytes, eof: true, client: 127.0.0.1, server: localhost, request: "GET /t HTTP/1.1", host: "localhost"
2023/11/21 17:14:07 [info] 1992222#0: *2 [proxy-wasm]["hostcalls" #1] [hostcalls] testing in "ResponseBody", client: 127.0.0.1, server: localhost, request: "GET /t HTTP/1.1", host: "localhost"
...                                                                                                                                                                                             
2023/11/21 17:14:07 [info] 1992222#0: *2 [proxy-wasm]["hostcalls" #1] property not found: wasmx.my_property, client: 127.0.0.1, server: localhost, request: "GET /t HTTP/1.1", host: "localhost"
```

I was trying to understand why this branch is being taken in `set_hpn_value` in this particular test case:

```c
    if (hpn->const_null) {
        return NGX_BUSY;
    }
```

(I did the rename to `const_null` myself because it still is not in the latest branch with code review).

Originally my goal was to find which tests were using this branch and I saw TEST 10, but this TEST 2 is concerning to me.

Reply:
Apparently something that I changed caused this behavior change, trying to find it. The original branch correctly shows `wasmx.my_property: my_value` for both `on_response_body` invocations, but since the test is not robust enough (see other comment), it did not catch the change when I made it. Now I can't find which change it was quite yet, since I committed it in a WIP commit.

Reply:
Something seems amiss; I've looked into this rabbit hole for some time, and now it seems the problematic change I made was to add a log line in the Lua C callback (`c_prop_getter`). Now it seems there actually is an issue with logging in Lua callbacks, for me this patch fails on a segfault most of the time (but sometimes passes):

```patch
diff --git a/t/04-openresty/ffi/304-proxy_wasm_set_property_getter.t b/t/04-openresty/ffi/304-proxy_wasm_set_property_getter.t
index e6c8f764..84922a1b 100644
--- a/t/04-openresty/ffi/304-proxy_wasm_set_property_getter.t
+++ b/t/04-openresty/ffi/304-proxy_wasm_set_property_getter.t
@@ -52,6 +52,7 @@ ok
 
 
 === TEST 2: set_property_getter() - setting host properties getter works on rewrite_by_lua
+--- ONLY
 --- wasm_modules: hostcalls
 --- http_config
     init_worker_by_lua_block {
@@ -74,9 +75,13 @@ ok
             assert(proxy_wasm.attach(_G.c_plan))
 
             local function getter(key)
+
+                print("here")
+
                 if key == "wasmx.my_property" then
                     return true, "my_value"
                 end
+
                 return false
             end
``` 

Backtrace:
```
#0  0x0000555cb8fd0f0a in log_wrapper (log=0x13e60800000067, ident=0x555cb90df8f9 "[lua] ", level=6, 
    L=0x7fcfc55f7d80) at ../ngx_lua-0.10.25/src/ngx_http_lua_log.c:104
#1  0x0000555cb8fd0ec0 in ngx_http_lua_print (L=0x7fcfc55f7d80)
    at ../ngx_lua-0.10.25/src/ngx_http_lua_log.c:86
#2  0x00007fcfc5759b93 in lj_BC_FUNCC ()
   from /code/ngx_wasm_module/work/buildroot/prefix/luajit/lib/libluajit-5.1.so.2
#3  0x0000555cb90ad7ef in ngx_proxy_wasm_properties_get (pwctx=0x555cbad04f50, path=0x7ffdc12fe650, 
    value=0x7ffdc12fe660)
    at /code/ngx_wasm_module/src/common/proxy_wasm/ngx_proxy_wasm_properties.c:1057
#4  0x0000555cb90a8a45 in ngx_proxy_wasm_hfuncs_get_property (instance=0x555cb9300900, 
    args=0x555cb9311000, rets=0x555cb9310fe0)
    at /code/ngx_wasm_module/src/common/proxy_wasm/ngx_proxy_wasm_host.c:779
#5  0x0000555cb909c39d in ngx_wavm_hfunc_trampoline (env=0x555cb93004c0, caller=0x7ffdc12fe868, 
    args=0x555cbad8f220, nargs=4, rets=0x555cbad8f280, nrets=1)
    at /code/ngx_wasm_module/src/wasm/vm/ngx_wavm_host.c:265
#6  0x00007fcfc49c3417 in wasmtime::func::c_callback_to_rust_fn::{{closure}}::hffefa239826578cc ()
   from /code/ngx_wasm_module/work/runtimes/wasmtime-14.0.3/lib/libwasmtime.so
#7  0x00007fcfc49ba82b in wasmtime::func::Func::invoke::h9cebbb299d0df8c4 ()
   from /code/ngx_wasm_module/work/runtimes/wasmtime-14.0.3/lib/libwasmtime.so
#8  0x00007fcfc49b7c64 in wasmtime::func::Caller<T>::with::h755a29576193b3ba ()
   from /code/ngx_wasm_module/work/runtimes/wasmtime-14.0.3/lib/libwasmtime.so
#9  0x00007fcfc496fba4 in wasmtime::trampoline::func::array_call_shim::h91a4e9d19e665282 ()
   from /code/ngx_wasm_module/work/runtimes/wasmtime-14.0.3/lib/libwasmtime.so
```

Reply:
I think overall the PR should also have tests of various Lua APIs being used within the callbacks; even if some APIs don't work, at least for circling their behavior.

Reply:
I'm able to reproduce this crash in `feat/ffi-properties-code-review` (same results as you), but not in `feat/ffi-properties-bridge-wip`. One suspicious difference is that in the bridge version, I'm calling `ngx_http_lua_set_req` explicitly to set the `r` of the `lua_State` before the callback runs. My commit log says this makes Valgrind happy, and I do recall seeing this type of error sometime during development, so I think that's how I solved this then.

Reply:
@thibaultcha This seems to solve the issue on my end. It's a Lua equivalent of `ngx_http_lua_set_req`, which I figured out by peeking at lua-resty-core:

```
diff --git lib/resty/wasmx/proxy_wasm.lua lib/resty/wasmx/proxy_wasm.lua
index 7934c4fd..6fecff3c 100644
--- lib/resty/wasmx/proxy_wasm.lua
+++ lib/resty/wasmx/proxy_wasm.lua
@@ -12,6 +12,7 @@ local pcall = pcall
 local ffi_gc = ffi.gc
 local ffi_new = ffi.new
 local ffi_str = ffi.string
+local getfenv = getfenv
 local tostring = tostring
 local subsystem = ngx.config.subsystem
 local get_request = wasmx.get_request
@@ -89,6 +90,16 @@ do
     -- We can't create one of these on each request,
     -- so we prepare these wrappers, which in turn call a Lua function.
 
+
+    local function set_r(r)
+        local pok, exdata = pcall(require, "thread.exdata")
+        if pok and exdata then
+            exdata(r)
+        else
+            getfenv(0).__ngx_req = r
+        end
+    end
+
     local function store_c_value(r, ckey, cvalue, lvalue, is_const)
         if lvalue ~= nil then
             local value = tostring(lvalue)
@@ -107,6 +118,7 @@ do
 
     c_prop_setter = ffi.cast("ngx_wasm_host_prop_fn_t", function(r, ckey,
                                                                  cvalue)
+        set_r(r)
         local lkey = ffi_str(ckey.data, ckey.len)
         local lval
         if cvalue.data ~= nil then
@@ -131,6 +143,7 @@ do
 
     c_prop_getter = ffi.cast("ngx_wasm_host_prop_fn_t", function(r, ckey,
                                                                  cvalue)
+        set_r(r)
         local lkey = ffi_str(ckey.data, ckey.len)
 
         local pok, ok, lvalue, is_const = pcall(lua_prop_getter, lkey)
```